wifiphisher/common/interfaces.py

```diff
@@ -319,233 +339,320 @@ def __init__(self):
-        :param jamming_argument: The jamming argument given by user
-        :param ap_argument: The AP argument given by user
         :type self: NetworkManager
-        :type jamming_argument: str
-        :type ap_argument: str
         :return: None
         :rtype: None
-        .. seealso:: NetworkAdapter
         """
 
-        # Setup the fields
-        self._interfaces = {}
-        self.ap_iface = ""
-        self.jam_iface = ""
+        self._name_to_object = dict()
+        self._active = set()
 
-        # Create, add and check compatibility for each interface
-        for interface in pyw.interfaces():
-            try:
-                self._interfaces[interface] = NetworkAdapter(interface)
-            except pyric.error as e:
-                pass
+    def is_interface_valid(self, interface_name, mode=None):
+        """
+        Check if interface is valid
+
+        :param self: A NetworkManager object
+        :param interface_name: Name of an interface
+        :param mode: The mode of the interface to be checked
+        :type self: NetworkManager
+        :type interface_name: str
+        :type mode: str
+        :return: True if interface is valid
+        :rtype: bool
+        .. note: The available modes are monitor and AP
+        """
+
+        # raise an error if interface can't be found
+        try:
+            interface_adapter = self._name_to_object[interface_name]
+
+            # raise an error if interface doesn't support the mode
+            if mode == "monitor" and not interface_adapter.has_monitor_mode:
+                raise InvalidInterfaceError(interface_name, mode)
+            elif mode == "AP" and not interface_adapter.has_ap_mode:
+                raise InvalidInterfaceError(interface_name, mode)
 
-    def up_ifaces(self, ifaces):
-        for i in ifaces:
-            card = pyw.getcard(i.get_name())
-            pyw.up(card)
-    
+            return True
+        except KeyError:
+            raise InvalidInterfaceError(interface_name)
 
-    def set_interface_mode(self, interface, mode):
+    def set_interface_mac(self, interface_name, mac_address):
         """
-        Set the desired mode to the network interface.
+        Set the specified MAC address for the interface
 
         :param self: A NetworkManager object
-        :param interface: A NetworkAdapter object
-        :param mode: The mode the interface should be set to
+        :param interface_name: Name of an interface
+        :param mac_address: A MAC address
         :type self: NetworkManager
-        :type interface: NetworkAdapter
+        :type interface_name: str
+        :type mac_address: str
+        :return: None
+        :rtype: None
+        .. note: This method will set the interface to managed mode
+        """
+
+        card = self._name_to_object[interface_name].card
+        self.set_interface_mode(interface_name, "managed")
+
+        try:
+            pyw.macset(card, mac_address)
```

Comment:
>Before setting mac address we should check whether the card is down. This is my first error to executing on this branch. I will fixed it on my local and see if there are any problems else.

Since I'm setting the card to `managed on line 395 the interface should b e 100% up.

>also checks if the card is in managed mode

On line 395 I'm setting it to `managed`.